lib/spargebra/src/parser.rs

```diff
@@ -9,64 +9,34 @@ use oxrdf::vocab::{rdf, xsd};
 use peg::parser;
 use peg::str::LineCol;
 use rand::random;
+use std::char;
 use std::collections::{HashMap, HashSet};
-use std::error::Error;
 use std::mem::take;
 use std::str::FromStr;
-use std::{char, fmt};
 
 /// Parses a SPARQL query with an optional base IRI to resolve relative IRIs in the query.
 pub fn parse_query(query: &str, base_iri: Option<&str>) -> Result<Query, ParseError> {
     let mut state = ParserState::from_base_iri(base_iri)?;
-    parser::QueryUnit(query, &mut state).map_err(|e| ParseError {
-        inner: ParseErrorKind::Parser(e),
-    })
+    parser::QueryUnit(query, &mut state).map_err(ParseError::Parser)
 }
 
 /// Parses a SPARQL update with an optional base IRI to resolve relative IRIs in the query.
 pub fn parse_update(update: &str, base_iri: Option<&str>) -> Result<Update, ParseError> {
     let mut state = ParserState::from_base_iri(base_iri)?;
-    let operations = parser::UpdateInit(update, &mut state).map_err(|e| ParseError {
-        inner: ParseErrorKind::Parser(e),
-    })?;
+    let operations = parser::UpdateInit(update, &mut state).map_err(ParseError::Parser)?;
     Ok(Update {
         operations,
         base_iri: state.base_iri,
     })
 }
 
 /// Error returned during SPARQL parsing.
-#[derive(Debug)]
-pub struct ParseError {
-    inner: ParseErrorKind,
-}
-
-#[derive(Debug)]
-enum ParseErrorKind {
-    InvalidBaseIri(IriParseError),
-    Parser(peg::error::ParseError<LineCol>),
-}
-
-impl fmt::Display for ParseError {
-    #[inline]
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        match &self.inner {
-            ParseErrorKind::InvalidBaseIri(e) => {
-                write!(f, "Invalid SPARQL base IRI provided: {e}")
-            }
-            ParseErrorKind::Parser(e) => e.fmt(f),
-        }
-    }
-}
-
-impl Error for ParseError {
-    #[inline]
-    fn source(&self) -> Option<&(dyn Error + 'static)> {
-        match &self.inner {
-            ParseErrorKind::InvalidBaseIri(e) => Some(e),
-            ParseErrorKind::Parser(e) => Some(e),
-        }
-    }
+#[derive(Debug, thiserror::Error)]
+pub enum ParseError {
+    #[error("Invalid SPARQL base IRI provided: {0}")]
+    InvalidBaseIri(#[from] IriParseError),
+    #[error(transparent)]
+    Parser(#[from] peg::error::ParseError<LineCol>),
```

Comment:
I would prefer to expose we use peg here for parsing. Maybe make variants private

Side note: I should maybe rename this type `SyntaxError` for consistency with the other crates

Reply:
I think it is actually very confusing to have so many identically named errors. I will add a prefix to keep them different